File: Java/프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;
class Solution {
    int[] dx = {0, 0, 1, -1};
    int[] dy = {1, -1, 0, 0};
    boolean[][] visited;
    int n, m, count;
    public int solution(int[][] land) {
        int answer = 0;
        n = land.length;
        m = land[0].length;
        visited = new boolean[n][m];
        int[] colSum = new int[m];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (land[j][i] == 0 || visited[j][i]) {
                    continue;
                }
                boolean[] visitCol = new boolean[m];
                count = 0;
                bfs(j, i, land, visitCol);
                for (int k = 0; k < m; k++) {
                    if (visitCol[k]) {
                        colSum[k] += count;
                    }
                }
            }
        }

        answer = Arrays.stream(colSum).max().getAsInt();

        return answer;
    }

    void bfs(int x, int y, int[][] land, boolean[] visitCol) {
        Deque<int[]> queue = new ArrayDeque<>();
        queue.add(new int[]{x, y});
        visited[x][y] = true;
        visitCol[y] = true;
        count++;

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();

            for (int i = 0; i < 4; i++) {
                int nx = cur[0] + dx[i];
                int ny = cur[1] + dy[i];
                if (nx < 0 || nx >= n || ny < 0 || ny >= m || land[nx][ny] == 0 || visited[nx][ny]) {
                    continue;
                }
                queue.add(new int[]{nx, ny});
                visited[nx][ny] = true;
                visitCol[ny] = true;
                count++;
            }
        }
    }
}

File: Python/프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
from collections import deque


def solution(land):
    n = len(land)
    m = len(land[0])

    col_sum = [0] * m
    visited = [[False] * m for _ in range(n)]
    dx = [0, 0, 1, -1]
    dy = [1, -1, 0, 0]

    for i in range(m):
        for j in range(n):
            if land[j][i] == 0 or visited[j][i]:
                continue

            q = deque([(j, i)])
            visited[j][i] = True
            count = 0
            columns = set()

            while q:
                cx, cy = q.popleft()
                count += 1
                columns.add(cy)

                for k in range(4):
                    nx = cx + dx[k]
                    ny = cy + dy[k]

                    if (
                        nx < 0
                        or nx >= n
                        or ny < 0
                        or ny >= m
                        or visited[nx][ny]
                        or land[nx][ny] == 0
                    ):
                        continue

                    q.append((nx, ny))
                    visited[nx][ny] = True

            for column in columns:
                col_sum[column] += count

    return max(col_sum)